[New] `forbid-component-props`: add `allowedForPatterns` and `disallowedForPatterns` options

Efimenko opened the pull request:
Initially, issue was described [here](https://github.com/jsx-eslint/eslint-plugin-react/issues/3686), but it was closed by [another PR](https://github.com/jsx-eslint/eslint-plugin-react/pull/3774) (although this PR implements different case).
So, this PR just reimplements `Regex` implementation with `globs` as was suggested in [one](https://github.com/jsx-eslint/eslint-plugin-react/issues/3686#issuecomment-1951461748) of the comments.

Efimenko commented:
@ljharb what do you think?

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -115,7 +142,9 @@ module.exports = {
       const prop = propName || propPattern;
       const options = {
         allowList: typeof value === 'string' ? [] : (value.allowedFor || []),
+        allowPatternList: typeof value === 'string' ? [] : value.allowedForPatterns || [],
         disallowList: typeof value === 'string' ? [] : (value.disallowedFor || []),
+        disallowPatternList: typeof value === 'string' ? [] : value.disallowedForPatterns || [],
```
```suggestion
        allowPatternList: [].concat(value.allowedForPatterns || []),
        disallowList: [].concat(value.disallowedFor || []),
        disallowPatternList: [].concat(value.disallowedForPatterns || []),
```
altho i'm not sure why any of these would be able to be strings - doesn't the schema prevent that?

Efimenko replied:
Yeah, you're right about schema, but there is a [default](https://github.com/jsx-eslint/eslint-plugin-react/pull/3805/files/40beeb52feaf2989486565a3df98f2415e1aa5d3#diff-95ccf9a24cad95dd24cdc5d59460de65a598a88ebe8c19a2ee6af25978549eb8L16) also for this rule, so that's why this code supports strings also.

ljharb replied:
I suppose the default could be made into an array, but nbd.

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -140,10 +169,48 @@ module.exports = {
         return false;
       }
 
+      function checkIsTagForbiddenByAllowList() {
+        return options.allowList.indexOf(tagName) === -1;
+      }
```
this seems overkill to make into a function, especially if it's used once. let's just inline it.

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -140,10 +169,48 @@ module.exports = {
+        return options.allowList.indexOf(tagName) === -1;
+      }
+
+      function checkIsTagForbiddenByAllowPatternList() {
+        if (options.allowPatternList.length === 0) {
+          return true;
+        }
+
+        return options.allowPatternList.every(
+          (pattern) => !minimatch(tagName, pattern)
+        );
+      }
```
kind of same here? i'd rather checkIsTagForbiddenByAllowOptions be mutiline with some early returns.

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -140,10 +169,48 @@ module.exports = {
+        return checkIsTagForbiddenByAllowList() && checkIsTagForbiddenByAllowPatternList();
+      }
+
+      function checkIsTagForbiddenByDisallowList() {
+        return options.disallowList.indexOf(tagName) !== -1;
+      }
+
+      function checkIsTagForbiddenByDisallowPatternList() {
+        if (options.disallowPatternList.length === 0) {
+          return false;
+        }
+
+        return options.disallowPatternList.some(
+          (pattern) => minimatch(tagName, pattern)
+        );
+      }
+
+      function checkIsTagForbiddenByDisallowOptions() {
+        return checkIsTagForbiddenByDisallowList() || checkIsTagForbiddenByDisallowPatternList();
+      }
```
same feedback on these three functions

Efimenko commented:
Honestly, not sure what to do with the "Automatic Rebase" action, didn't find any info in [CONTRIBUTING.md](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/CONTRIBUTING.md)

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -66,12 +71,20 @@ module.exports = {
                   message: { type: 'string' },
                 },
-                required: ['disallowedFor'],
+                anyOf: [
```
should this be:
```suggestion
                oneOf: [
```
?

Efimenko replied:
Not sure, because with `oneOf` you can't use `disallowedFor` and `disallowedForPatterns` simultaneously, while it is still a case. Here is one of the test cases for example: https://github.com/jsx-eslint/eslint-plugin-react/pull/3805/files#diff-0095adf2f1f3e139bb1a1666d7611671ad90ae66f67de0f5cdc7cb8884351968R825-R847

Efimenko replied:
@ljharb so, based on the documentation, only one can be successful at a time https://json-schema.org/draft/2020-12/json-schema-core#section-10.2.1.3-2

ljharb replied:
sure. but "only one is required" is the intended semantics, right?

i suppose they're identical semantics, but i read "anyOf" with more optionality than "oneOf", personally ¯\\\_(ツ)_/¯

Efimenko replied:
Yeah, semantically about "at least one required" they are identical probably, but "use several of them simultaneously" is not.

ljharb replied:
right - and "any of" implies multiple matches are expected, "one of" implies only one match is expected

Efimenko replied:
So, to summarize:
- `anyOf` gives the ability to use several fields like `disallowedFor` and `disallowedForPatterns` at the same time;
- `oneOf` forbids the use of several fields at the same time;

So, here we are, just discussing the difference, or we're considering the replacement with `oneOf`? A bit hard to understand 😄

ljharb replied:
lol i'm saying that i think that the actual behavior will be identical with anyOf or oneOf here, but that `oneOf` seems more appropriate to me.

I won't hold up the PR further on this topic, though.

Efimenko replied:
I got it, so, actually, no.
With `oneOf`, if you pass at the same time `disallowedFor` and `disallowedForPatterns ` - it will be a validation error.
With `anyOf` - it won't.

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -95,9 +113,18 @@ module.exports = {
                   message: { type: 'string' },
                 },
-                required: ['disallowedFor'],
+                anyOf: [
```
```suggestion
                oneOf: [
```
this one too?

Efimenko replied:
the same https://github.com/jsx-eslint/eslint-plugin-react/pull/3805#discussion_r1773456872

Efimenko commented:
> Should there also be examples of using `propNamePatterns` with one of these new options?

I've added one: https://github.com/jsx-eslint/eslint-plugin-react/pull/3805/files#diff-1cf5e390e9444274ee8064c64a72965facfa2171c5453a130f0329e57fe991adR79-R83

Also, I've added one for a more complex setup: https://github.com/jsx-eslint/eslint-plugin-react/pull/3805/files#diff-1cf5e390e9444274ee8064c64a72965facfa2171c5453a130f0329e57fe991adR106-R113 (Relates to this comment https://github.com/jsx-eslint/eslint-plugin-react/pull/3805#discussion_r1773456872)

Efimenko commented:
For the future, correct me if I'm wrong, but here, `rebase` is a preferable option over `merge`.

ljharb commented:
Yes, feel free to do whatever you like, but I'll rebase and clean it up as part of landing it.
